sklearn/ensemble/voting_classifier.py

```diff
@@ -0,0 +1,217 @@
+"""
+Soft Voting/Majority Rule classifier.
+
+This module contains a Soft Voting/Majority Rule classifier for
+classification estimators.
+
+"""
+
+# Authors: Sebastian Raschka <se.raschka@gmail.com>
+#
+# Licence: BSD 3 clause
+
+from ..base import BaseEstimator
+from ..base import ClassifierMixin
+from ..base import TransformerMixin
+from ..base import clone
+from ..preprocessing import LabelEncoder
+from ..externals import six
+from ..pipeline import _name_estimators
+
+from collections import defaultdict
+import numpy as np
+import operator
+
+
+class VotingClassifier(BaseEstimator, ClassifierMixin, TransformerMixin):
+
+    """ Soft Voting/Majority Rule classifier for unfitted estimators.
+
+    Parameters
+    ----------
+    estimators : list of (string, estimator) tuples
+      Invoking the `fit` method on the `VotingClassifier` will fit clones
+      of those original estimators that will be stored in the class attribute
+      `self.estimators_`.
+
+    voting : str, {'hard', 'soft'} (default='hard')
+      If 'hard', uses predicted class labels for majority rule voting.
+      Else if 'soft', predicts the class label based on the argmax of
+      the sums of the predicted probalities, which is recommended for
+      an ensemble of well-calibrated classifiers.
+
+    weights : array-like, shape = [n_classifiers], optional (default=`None`)
+      Sequence of weights (`float` or `int`) to weight the occurances of
+      predicted class labels (`hard` voting) or class probabilities
+      before averaging (`soft` voting). Uses uniform weights if `None`.
+
+    Attributes
+    ----------
+    classes_ : array-like, shape = [n_predictions]
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn.linear_model import LogisticRegression
+    >>> from sklearn.naive_bayes import GaussianNB
+    >>> from sklearn.ensemble import RandomForestClassifier
+    >>> clf1 = LogisticRegression(random_state=1)
+    >>> clf2 = RandomForestClassifier(random_state=1)
+    >>> clf3 = GaussianNB()
+    >>> X = np.array([[-1, -1], [-2, -1], [-3, -2], [1, 1], [2, 1], [3, 2]])
+    >>> y = np.array([1, 1, 1, 2, 2, 2])
+    >>> eclf1 = VotingClassifier(estimators=[
+    ...         ('lr', clf1), ('rf', clf2), ('gnb', clf3)], voting='hard')
+    >>> eclf1 = eclf1.fit(X, y)
+    >>> print(eclf1.predict(X))
+    [1 1 1 2 2 2]
+    >>> eclf2 = VotingClassifier(estimators=[
+    ...         ('lr', clf1), ('rf', clf2), ('gnb', clf3)],
+    ...         voting='soft')
+    >>> eclf2 = eclf2.fit(X, y)
+    >>> print(eclf2.predict(X))
+    [1 1 1 2 2 2]
+    >>> eclf3 = VotingClassifier(estimators=[
+    ...        ('lr', clf1), ('rf', clf2), ('gnb', clf3)],
+    ...        voting='soft', weights=[2,1,1])
+    >>> eclf3 = eclf3.fit(X, y)
+    >>> print(eclf3.predict(X))
+    [1 1 1 2 2 2]
+    >>>
+    """
+
+    def __init__(self, estimators, voting='hard', weights=None):
```

Comment:
why do we default to hard voting again? Because the classifiers will not be calibrated against each other? [if this is the reason, should we document it?]